src/core/toPath.js

```diff
@@ -102,139 +102,89 @@ const allowingArrays = fn => arg => {
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-          } else { // If a slice quote is found
+const splitAtFirstOccurence = (str, separators) => {
+  const partitionIndex = separators
+    .map(separator => str.indexOf(separator))
+    .filter(index => index >= 0)
+    .reduce((minIndex, index) => Math.min(minIndex, index), str.length)
+  return [str.substring(0, partitionIndex), str.substr(partitionIndex, 1), str.substring(partitionIndex + 1)]
+}
 
-            // Fetch slice start and end, and parse them as slice indexes (empty or valid int)
-            const sliceStart = arrayIndexValue.substring(0, sliceDelimIndex), sliceEnd = arrayIndexValue.substring(sliceDelimIndex + 1)
-            const nSliceStart = toSliceIndex(sliceStart), nSliceEnd = toSliceIndex(sliceEnd)
+/**
+ * 
+ * @param {string} str string to parse, expected to start with an opening square bracket followed by a quote char
+ * @param {string} quote the quote char
+ * @returns {[string, string]} a tuple of the dequoted path segment and the rest of the input string
+ * @example parseQuotedBracketNotation('["abc"].def', '"') // ['abc', 'def']
+ * @example parseQuotedBracketNotation('["abc', '"') // ['abc', '']
+ * @example parseQuotedBracketNotation('abc', '"') // ['c', '']
+ * @example parseQuotedBracketNotation('ab', '"') // [undefined, '']
+ */
+const parseQuotedBracketNotation = (str, quote) => {
+  const [match, prop, remainingStr] =
+    str.match(new RegExp(`^\\[${quote}(.*?[^\\\\])${quote}\\]?\\.?(.*)$`)) || []
+  if (!match)
+    return [str.substring(2) || undefined, '']
+  return [unescapeQuotes(prop, quote), remainingStr]
+}
 
-            // Add array index to path, as a slice if both slice indexes are valid (undefined or int), or as a string
-            path.push(isSliceIndex(nSliceStart) && isSliceIndex(nSliceEnd) ? [nSliceStart, nSliceEnd] : arrayIndexValue)
-          }
-        }
+/**
+ * @typedef {number|undefined} SliceIndex
+ * @typedef {[SliceIndex,SliceIndex]} Slice
+ * @param {string} str string to parse, expected to start with an opening square bracket
+ * @returns {[(string|number|Slice),string]} a tuple of the path segment and the rest of the input string
+ * @example parseBareBracketNotation('[123].def') // [123, 'def']
+ * @example parseBareBracketNotation('[123]def') // [123, 'def']
+ * @example parseBareBracketNotation('[12a].def') // ['12a', 'def']
+ * @example parseBareBracketNotation('[3.4].def') // ['3.4', 'def']
+ * @example parseBareBracketNotation('[:].def') // [[undefined, undefined], 'def']
+ * @example parseBareBracketNotation('[14:].def') // [[14, undefined], 'def']
+ * @example parseBareBracketNotation('[:190].def') // [[undefined, 190], 'def']
+ * @example parseBareBracketNotation('[14:190].def') // [[14, 190], 'def']
+ * @example parseBareBracketNotation('[14:190.def') // ['14:190.def', '']
```

Comment:
These examples should probably be turned into tests. The two below maybe not, I don't think we want to test what basically is undefined behavior (the function should not be called with these arguments).